rep-2004.rst

```diff
@@ -0,0 +1,808 @@
+* Quality Level 4:
+
+  * demos, tutorials, and experiments
+  * e.g. research packages, ``demo_nodes_cpp``, ``examples_rclcpp_minimal_publisher``, etc.
+
+* Quality Level 5:
+
+  * default quality level
+  * e.g. for code that fails to meet requirements in 'Level 4'
+
+While each quality level will have different requirements, it's always possible to overachieve in certain requirements even if other requirements prevent a package from moving up to the next quality level.
+
+Quality Level 1
+^^^^^^^^^^^^^^^
+
+This category should be used for packages which are required for a reasonable ROS system in a production environment.
+These are the packages that remain as requirements for a basic ROS system to run after you remove development tools, build tools, and introspection tools.
+
+Packages that would not normally fit this description can still be called 'Level 1'.
+If there is a need for a particular package in a reasonable production scenario, then that package should be considered for this category as well.
+However, packages which we consider essential to getting a robot up and running quickly, but perhaps provide a generic solution to a problem should probably not start out as 'Level 1' due to the high effort required to get a package to 'Level 1' and maintain it there.
+
+For example, packages which provide intra-process communication, inter-process communication, generated message runtime code, node lifecycle, etc. should probably all be considered for 'Level 1'.
+However, a package which provides pose estimation (like ``robot_pose_ekf``\ ) is a generic solution for something that most people need, but is often replaced with a domain specific solution in production, and therefore it should probably not start out as 'Level 1'.
+Such a package may upgrade to 'Level 1' at a later date, if it proves to be a solution that people want to use in their products.
+
+Tools, like ``rostopic``\ , generally do not fall into this category either, but are not categorically excluded.
+For example, it may be the case that the tool which launches and verifies a ROS graph (``ros2launch``\ ) may need to be considered 'Level 1' for use in production systems.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+Requirements to be considered a 'Level 1' package:
+
+.. _Version Policy:
+
+1. **Version Policy:**
+
+   .. _1.i:
+
+   i. Must have a version policy (e.g. ``semver``)
+
+   .. _1.ii:
+
+   ii. Must be at a stable version (e.g. for ``semver`` that means version >= 1.0.0)
+
+   .. _1.iii:
+
+   iii. Must have a strictly declared public API
+
+   .. _1.iv:
+
+   iv. Must have a policy for API stability
+
+   .. _1.v:
+
+   v. Must have a policy for ABI stability
+
+   .. _1.vi:
+
+   vi. Must have a policy that keeps API and ABI stability within a released ROS distribution
+
+.. _Change Control Process:
+
+2. **Change Control Process:**
+
+   .. _2.i:
+
+   i. Must have all code changes occur through a change request (e.g. pull request, merge request, etc.)
```

Comment:
We need to define a time window for this criteria. E.g. Level 1 requires that all pushes since the last 30 days? last release? to have all been reviewed. Otherwise, it feels like if a single PR skips review, the repo will never be level 1 ever again.

Reply:
This is a requirement of the process, not the code's history. In my opinion this simply means you need to have a policy that says you do this. It doesn't mean that if you ever in your past did not have this policy you cannot be level 1.

Obviously you can argue that code that at one point didn't have this policy in place isn't "good enough", but I think that's also impractical since basically all code in our ecosystem may have had that in their past.

Instead, I would say, when transitioning to requiring code reviews you have to do a review of the entire code base, or other justify why that is not needed, in order to "catch things up". I don't think saying, "wait 30 days and what you did before then doesn't matter any more" doesn't really improve the code.

Reply:
Sure, but if you take a level 1 package, and apply a change without a review, what is the consequence?

Reply:
Or maybe we could make this more actionable?

Taking GitHub as an example:

* Require pull request reviews before merging
* Require status checks to pass before merging
* Require branches to be up to date before merging
* DCO

Reply:
> if you take a level 1 package, and apply a change without a review, what is the consequence?

I'd say that it's up to each project to decide and document their policies.

Since this REP has a few "For ROS 2 this means...", I think we could come up with a policy for ROS 2 core and document it here or in the developer guide. I think the guide is a better place for *all* these ROS 2 specific guidelines, so we don't need to come back and change the REP every time the technology evolves.

So, talking about ROS 2:

> define a time window for this criteria

I don't think a time-based "penalty" makes sense in this case. The project could revert the changes and be back at where they were straight away.

> maybe we could make this more actionable

I think these are all good things for the ROS 2 core to adopt and document in the developer guide. One major technical difficulty right now with adding GitHub merge checks is that the bloom release process needs to commit directly to the release branch. The team is looking into ways around this for the core packages, possibly using GitHub bots.